test/34.database.table.js

```diff
@@ -0,0 +1,593 @@
+			.to.throw(Database.SqliteError);
+	});
+	it('should throw an exception if the database is busy', function () {
+		let ranOnce = false;
+		for (const x of this.db.prepare('SELECT 2').pluck().iterate()) {
+			expect(x).to.equal(2);
+			ranOnce = true;
+			expect(() => this.db.table('a', { columns: ['x'], rows: function* () {} })).to.throw(TypeError);
+		}
+		expect(ranOnce).to.be.true;
+		this.db.table('b', { columns: ['x'], rows: function* () {} });
+	});
+	it('should cause the database to become busy when querying the virtual table', function () {
+		let checkCount = 0;
+		const expectBusy = function* () {
+			for (let i = 0; i < 3; ++i) {
+				expect(() => this.db.exec('SELECT * FROM a')).to.throw(TypeError);
+				expect(() => this.db.prepare('SELECT 555')).to.throw(TypeError);
+				expect(() => this.db.pragma('cache_size')).to.throw(TypeError);
+				expect(() => this.db.function('x', () => {})).to.throw(TypeError);
+				expect(() => this.db.table('y', { columns: ['x'], rows: function* () {} })).to.throw(TypeError);
+				checkCount += 1;
+				yield [i];
+			}
+		};
+		this.db.table('a', { columns: ['x'], rows: function* () {} });
+		this.db.table('b', { columns: ['x'], rows: expectBusy });
+
+		expect(this.db.prepare('SELECT * FROM b').pluck().all()).to.deep.equal([0, 1, 2]);
+		expect(checkCount).to.equal(3);
+
+		this.db.exec('SELECT * FROM a');
+		this.db.prepare('SELECT 555');
+		this.db.pragma('cache_size');
+		this.db.function('xx', () => {});
+		this.db.table('yy', { columns: ['x'], rows: function* () {} })
+	});
+	it('should cause the virtual table to throw when yielding an invalid value', function () {
+		this.db.table('a', {
+			columns: ['x'],
+			*rows() { yield [42]; }
+		});
+		this.db.table('b', {
+			columns: ['x'],
+			*rows() { yield 42; }
+		});
+		this.db.table('c', {
+			columns: ['x'],
+			*rows() { yield; }
+		});
+		this.db.table('d', {
+			columns: ['x'],
+			*rows() { yield null; }
+		});
+		expect(this.db.prepare('SELECT * FROM a').get()).to.deep.equal({ x: 42 });
+		expect(() => this.db.prepare('SELECT * FROM b').get()).to.throw(TypeError);
+		expect(() => this.db.prepare('SELECT * FROM c').get()).to.throw(TypeError);
+		expect(() => this.db.prepare('SELECT * FROM d').get()).to.throw(TypeError);
+	});
+	it('should allow arrays to be yielded as rows', function () {
+		const rows = [
+			{ a: null, b: 123, c: 456.789, d: 'foo', e: Buffer.from('bar') },
+			{ a: null, b: 987, c: 654.321, d: 'oof', e: Buffer.from('rab') },
+		];
+		this.db.table('vtab', {
+			columns: ['a', 'b', 'c', 'd', 'e'],
+			*rows() {
+				for (const obj of rows) {
+					yield Object.values(obj);
+				}
+			},
+		});
+		expect(this.db.prepare('SELECT * FROM vtab').all()).to.deep.equal(rows);
+	});
+	it('should allow objects to be yielded as rows', function () {
+		const rows = [
+			{ a: null, b: 123, c: 456.789, d: 'foo', e: Buffer.from('bar') },
+			{ a: null, b: 987, c: 654.321, d: 'oof', e: Buffer.from('rab') },
```

Comment:
Maybe add a row with reversed keys and one with shuffled keys. Since `extractRowObject` will iterate in source order I originally thought this could cause an issue with filling the array with `null`. There was no issue but the test should make sure the order of the object keys does not matter.

Reply:
Good idea, adding...